File: Section06OptionalChallenges/DivideAndConquer_sortedFrequency.js
function sortedFrequency(arr, num) {
  let firstIndex = findFirst(arr, num);
  if (firstIndex < 0) {
    return -1;
  }
  let lastIndex = findLast(arr, num);
  return lastIndex - firstIndex + 1;
}

function findFirst(arr, num, lowIndex = 0, highIndex = arr.length - 1) {
  if (highIndex >= lowIndex) {
    let midIndex = Math.floor((highIndex + lowIndex) / 2);
    if (arr[midIndex] === num && (midIndex === 0 || arr[midIndex - 1] < num)) {
      return midIndex;
    } else if (arr[midIndex] < num) {
      return findFirst(arr, num, midIndex + 1, highIndex);
    } else {
      return findFirst(arr, num, lowIndex, midIndex - 1);
    }
  }
  return -1;
}

function findLast(arr, num, lowIndex = 0, highIndex = arr.length - 1) {
  if (highIndex >= lowIndex) {
    let midIndex = Math.floor((highIndex + lowIndex) / 2);
    if (arr[midIndex] === num && (midIndex === arr.length - 1 || arr[midIndex + 1] > num)) {
      return midIndex;
    } else if (arr[midIndex] > num) {
      return findLast(arr, num, lowIndex, midIndex - 1);
    } else {
      return findLast(arr, num, midIndex + 1, highIndex);
    }
  }
  return -1;
}

console.log(sortedFrequency([1,1,2,2,2,2,3],2));
console.log(sortedFrequency([1,1,2,2,2,2,3],3));
console.log(sortedFrequency([1,1,2,2,2,2,3],1));
console.log(sortedFrequency([1,1,2,2,2,2,3],4));

File: Section06OptionalChallenges/Section06OptionalChallenges_DOTNET/Section06OptionalChallenges/DivideAndConquer_SortedFrequency.cs
using System;

namespace Section06OptionalChallenges.DivideAndConquer;

public partial class DivideAndConquer
{
  public static int SortedFrequency(int[] arr, int num)
  {
    var firstIndex = FindFirst(arr, num, 0, arr.Length -1);
    if (firstIndex < 0)
    {
      return -1;
    }
    var lastIndex = FindLast(arr, num, 0, arr.Length - 1);
    return lastIndex - firstIndex + 1;
  }

  private static int FindFirst(int[] arr, int num, int lowIndex, int highIndex)
  {
    if (highIndex >= lowIndex)
    {
      var midIndex = (int)Math.Floor((double)(highIndex + lowIndex) / 2);
      if (arr[midIndex] == num && (midIndex == 0 || arr[midIndex - 1] < num))
      {
        return midIndex;
      }
      else if (arr[midIndex] < num)
      {
        return FindFirst(arr, num, midIndex + 1, highIndex);
      }
      else
      {
        return FindFirst(arr, num, lowIndex, midIndex - 1);
      }
    }
    return -1;
  }

  private static int FindLast(int[] arr, int num, int lowIndex, int highIndex)
  {
    if (highIndex >= lowIndex)
    {
      var midIndex = (int)Math.Floor((double)(highIndex + lowIndex) / 2);
      if (arr[midIndex] == num && (midIndex == arr.Length - 1 || arr[midIndex + 1] > num))
      {
        return midIndex;
      }
      else if (arr[midIndex] > num)
      {
        return FindLast(arr, num, lowIndex, midIndex - 1);
      }
      else
      {
        return FindLast(arr, num, midIndex + 1, highIndex);
      }
    }
    return -1;
  }
}